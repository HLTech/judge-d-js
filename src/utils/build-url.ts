import { URL } from 'url';

export function buildUrl(
    url: string,
    serviceName: string,
    serviceVersion: string
) {
    return new URL(
        `./contracts/services/${serviceName}/versions/${serviceVersion}`,
        url
    ).toString();
}
