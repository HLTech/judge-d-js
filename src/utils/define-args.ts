import yargs from 'yargs';

export interface CliArguments {
    url: string;
    pactsDir: string;
    serviceName: string;
    serviceVersion: string;
}

export function defineArgs(cliArgs: string[]): yargs.Arguments<CliArguments> {
    return yargs(cliArgs)
        .command<CliArguments>('publish', 'Publish contracts', (yargs) => {
            yargs.options({
                url: {
                    type: 'string',
                    demandOption: true,
                    describe: 'Url to judge-d instance',
                },
                pactsDir: {
                    type: 'string',
                    demandOption: true,
                    describe: 'Path to directory with pacts',
                },
                serviceName: {
                    type: 'string',
                    demandOption: true,
                    describe: 'Service name',
                },
                serviceVersion: {
                    type: 'string',
                    demandOption: true,
                    describe: 'Service version',
                },
            });
        })
        .strict()
        .demandCommand(1, 'You need at least one command before moving on')
        .help().argv;
}
