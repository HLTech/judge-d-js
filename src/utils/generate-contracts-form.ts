import { Pact } from './read-pacts';

export function generateContractsForm(pactFiles: Pact[]): ServiceContractsForm {
    return pactFiles.reduce<ServiceContractsForm>(
        (serviceContractsForm, fileContent) => {
            if (fileContent.provider?.name) {
                return {
                    ...serviceContractsForm,
                    expectations: {
                        ...serviceContractsForm.expectations,
                        [fileContent.provider.name]: {
                            rest: {
                                value: JSON.stringify(fileContent),
                                mimeType: 'application-json',
                            },
                        },
                    },
                };
            }
            return serviceContractsForm;
        },
        {
            capabilities: {},
            expectations: {},
        }
    );
}

export interface ServiceContractsForm {
    capabilities: Record<string, RestContract>;
    expectations: Record<string, RestContract>;
}

interface RestContract {
    rest: {
        value: string;
        mimeType: 'application-json';
    };
}
