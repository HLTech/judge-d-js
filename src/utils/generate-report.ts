import ejs from 'ejs';
import path from 'path';

type Status = 'OK' | 'FAILED';

export interface ValidationResult {
    validationStatus?: Status;
    consumerAndProvider: {
        providerName: string;
        providerVersion: string;
        consumerName: string;
        consumerVersion: string;
    };
    interactions: Array<{
        validationResult: Status;
        interactionName: string;
        errors: string[];
    }>;
}

export async function generateReport(validationResults?: ValidationResult[]) {
    const pathToTemplate = path.resolve(
        __dirname,
        '../../template/report-template.ejs'
    );
    return ejs.renderFile(pathToTemplate, {
        validationResults,
    });
}
