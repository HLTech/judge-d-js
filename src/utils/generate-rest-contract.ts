import { RestContract } from '../types';

export function generateRestContract(value: object): RestContract {
    return {
        rest: {
            value: JSON.stringify(value),
            mimeType: 'application/json',
        },
    };
}
