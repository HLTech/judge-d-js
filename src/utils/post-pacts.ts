import { ServiceContractsForm } from './generate-contracts-form';
import axios from 'axios';

export function postPacts(url: string, pacts: ServiceContractsForm) {
    return axios.post(url, pacts);
}
