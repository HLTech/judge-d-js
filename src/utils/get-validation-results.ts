import axios from 'axios';
import { CliVerifyArguments } from './define-args';
import { ValidationResult } from './generate-report';
import { stringify } from 'querystring';
import { URL } from 'url';

export async function getValidationResults(argv: CliVerifyArguments) {
    const { serviceName, url, serviceVersion, environment } = argv;

    const query = stringify({ environment });
    const judgeDUrl = new URL(
        `./environment-compatibility-report/${serviceName}:${serviceVersion}?${query}`,
        url
    ).toString();

    const response = await axios.get<ValidationResult[]>(judgeDUrl);
    return response.data;
}
