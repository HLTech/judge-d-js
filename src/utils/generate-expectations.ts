import { Pact, ServiceContractsForm } from '../types';
import { generateRestContract } from './generate-rest-contract';

export function generateExpectations(
    pactFiles: Pact[]
): ServiceContractsForm['expectations'] {
    const expectations: ServiceContractsForm['expectations'] = {};

    for (const pactFile of pactFiles) {
        if (pactFile.provider?.name) {
            expectations[pactFile.provider.name] = generateRestContract(
                pactFile
            );
        }
    }

    return expectations;
}
