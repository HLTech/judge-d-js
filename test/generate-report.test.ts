import { generateReport } from '../src/utils/generate-report';
import { validationResultMockFactory } from './mocks/validationResult.mock';

describe('generateReport', () => {
    test('returns html report with all interactions passed', async () => {
        const validationResultsMock = [validationResultMockFactory.build()];

        const report = await generateReport(validationResultsMock);

        expect(report).toMatchSnapshot();
    });

    test('returns html report with failed interaction', async () => {
        const validationResultsMock = [
            validationResultMockFactory.build({
                interactions: [
                    {
                        validationResult: 'OK',
                        interactionName: 'interaction name',
                        errors: [],
                    },
                    {
                        validationResult: 'FAILED',
                        interactionName: 'failed interaction',
                        errors: ['Provider can not be found.'],
                    },
                ],
            }),
        ];

        const report = await generateReport(validationResultsMock);

        expect(report).toMatchSnapshot();
    });
});
