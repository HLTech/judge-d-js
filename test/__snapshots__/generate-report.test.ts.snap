// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateReport returns html report with all interactions passed 1`] = `
"<!DOCTYPE html>
<html>
<head>
    <title>Pact validation report</title>

    <link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.1.0/css/bootstrap.min.css\\"
          integrity=\\"sha384-9gVQ4dYFwwWSjIDZnLEWnxCjeSWFphJiwGPXr1jddIhOegiu1FwO5qRGvFXOdJZ4\\" crossorigin=\\"anonymous\\">
</head>
<body>

<div class=\\"container\\">
    <h1>Pact validation result</h1>
    
        <ul class=\\"list-group list-group-flush\\">
            
                <li class=\\"list-group-item\\">
                    
                        <h5>
                            Interaction between
                        </h5>
                        <h5>
                            consumer: consumer-name
                        </h5>
                        <h5>
                            and
                        </h5>
                        <h5>
                            provider: provider-name
                        </h5>
                        <br/>
                        Consumer version = consumer-version<br/>
                        Provider version = provider-version
                        <br/>
                        <br/>
                        <table class=\\"table table-sm\\">
                            <caption>interactions</caption>
                            <thead class=\\"thead-dark\\">
                            <tr>
                                <th scope=\\"col\\">Interaction name</th>
                                <th scope=\\"col\\">Result</th>
                                <th scope=\\"col\\">Errors</th>
                            </tr>
                            </thead>
                            <tbody>
                            
                                
                                    
                                
                                <tr class=table-success>
                                    <td>interaction name</td>
                                    <td>OK</td>
                                    <td>
                                        
                                            -
                                        
                                    </td>
                                </tr>
                            
                                
                                    
                                
                                <tr class=table-success>
                                    <td>second interaction</td>
                                    <td>OK</td>
                                    <td>
                                        
                                            -
                                        
                                    </td>
                                </tr>
                            
                            <tbody>
                        </table>
                    
                </li>
            
        </ul>
    
</div>

</body>
</html>
"
`;

exports[`generateReport returns html report with failed interaction 1`] = `
"<!DOCTYPE html>
<html>
<head>
    <title>Pact validation report</title>

    <link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.1.0/css/bootstrap.min.css\\"
          integrity=\\"sha384-9gVQ4dYFwwWSjIDZnLEWnxCjeSWFphJiwGPXr1jddIhOegiu1FwO5qRGvFXOdJZ4\\" crossorigin=\\"anonymous\\">
</head>
<body>

<div class=\\"container\\">
    <h1>Pact validation result</h1>
    
        <ul class=\\"list-group list-group-flush\\">
            
                <li class=\\"list-group-item\\">
                    
                        <h5>
                            Interaction between
                        </h5>
                        <h5>
                            consumer: consumer-name
                        </h5>
                        <h5>
                            and
                        </h5>
                        <h5>
                            provider: provider-name
                        </h5>
                        <br/>
                        Consumer version = consumer-version<br/>
                        Provider version = provider-version
                        <br/>
                        <br/>
                        <table class=\\"table table-sm\\">
                            <caption>interactions</caption>
                            <thead class=\\"thead-dark\\">
                            <tr>
                                <th scope=\\"col\\">Interaction name</th>
                                <th scope=\\"col\\">Result</th>
                                <th scope=\\"col\\">Errors</th>
                            </tr>
                            </thead>
                            <tbody>
                            
                                
                                    
                                
                                <tr class=table-success>
                                    <td>interaction name</td>
                                    <td>OK</td>
                                    <td>
                                        
                                            -
                                        
                                    </td>
                                </tr>
                            
                                
                                    
                                
                                <tr class=table-danger>
                                    <td>failed interaction</td>
                                    <td>FAILED</td>
                                    <td>
                                        
                                            
                                                Provider can not be found. <br/>
                                            
                                        
                                    </td>
                                </tr>
                            
                            <tbody>
                        </table>
                    
                </li>
            
        </ul>
    
</div>

</body>
</html>
"
`;
